src/main/java/mate/academy/dao/BookDaoImpl.java

```diff
@@ -0,0 +1,110 @@
+            }
+            ResultSet generatedKeys = statement.getGeneratedKeys();
+            if (generatedKeys.next()) {
+                Long id = generatedKeys.getObject(1, Long.class);
+                book.setId(id);
+            }
+            return book;
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't add new book to the DB", e);
+        }
+    }
+
+    @Override
+    public Optional<Book> findById(Long id) {
+        try (Connection connection = ConnectionUtil.getConnection();
+                 PreparedStatement statement = connection.prepareStatement(FIND_BY_ID_QUERY)) {
+            statement.setLong(1, id);
+            ResultSet resultSet = statement.executeQuery();
+            if (resultSet.next()) {
+                return Optional.of(mapToBook(resultSet));
+            }
+            return Optional.empty();
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't find book by id = " + id, e);
+        }
+    }
+
+    @Override
+    public List<Book> findAll() {
+        List<Book> books = new ArrayList<>();
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(FIND_ALL_QUERY)) {
+            ResultSet resultSet = statement.executeQuery();
+            while (resultSet.next()) {
+                books.add(mapToBook(resultSet));
+            }
+            return books;
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't find list of books", e);
+        }
+    }
+
+    @Override
+    public Book update(Book book) {
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(UPDATE_QUERY)) {
+            statement.setString(1, book.getTitle());
+            statement.setBigDecimal(2, book.getPrice());
+            statement.setLong(3, book.getId());
+            if (statement.executeUpdate() < 1) {
+                throw new SQLException("Expected to insert at least 1 row, but inserted 0 rows.");
+            }
+            return book;
+        } catch (SQLException e) {
+            throw new DataProcessingException(
+              "Can't update information in the row with id = " + book.getId(), e);
+        }
+    }
+
+    @Override
+    public boolean deleteById(Long id) {
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(DELETE_QUERY)) {
+            statement.setLong(1, id);
+            return statement.executeUpdate() > 0;
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't delete book with id = " + id, e);
+        }
+    }
+
+    private Book mapToBook(ResultSet resultSet) throws SQLException {
+        return new Book(
+            resultSet.getLong("id"),
```

Comment:
pay attention to what is said about this method in the checklist

Reply:
I don't get it. In the checklist it's said that I should use constructor OR setters (but not both of them). So I use constructor. Or do I need to create variables for all the fields and only then put it in constructor?

Reply:
@Alexsandr-Yablunovskyi the problem here is that `getLong` will return 0 if id is absent, which you should omit

therefore you should use `getObject("id", Long.class)` so that you'll get null in case id is absent

alternatively you can handle such scenario during creation by throwing exception when id is absent